src/macaron/parsers/bashparser.py

```diff
@@ -115,78 +147,109 @@ def parse(bash_content: str, macaron_path: str = "") -> dict:
-
-    parsed_parent = parse(bash_content)
-    caller_commands = parsed_parent.get("commands", [])
-    if caller_commands:
-        yield BashCommands(
-            caller_path=ci_file,
-            CI_path=ci_file,
-            CI_type=ci_type,
-            commands=caller_commands,
-            job_name=job_name,
-            step_name=step_name,
-        )
+    if recursion_depth > defaults.getint("bashparser", "recursion_depth", fallback=3):
+        raise CallGraphError(f"The analysis has reached maximum recursion depth {recursion_depth} at {source_path}.")
+    parsed_content = {}
+    working_dir = None
+    match node_type:
+        case BashScriptType.INLINE:
+            if parsed_obj is None:
+                raise CallGraphError(f"Unable to find the parsed AST for the CI step at {source_path}.")
+            step_exec = validate_step(parsed_obj)
+            if step_exec is None:
+                raise CallGraphError(f"Unable to validate parsed AST for the CI step at {source_path}.")
+
+            working_dir = step_exec.get("WorkingDirectory")
+            run_script = validate_run_step(parsed_obj)
+            if run_script is None:
+                raise CallGraphError(f"Invalid run step at {source_path}.")
+            parsed_content = parse(run_script, macaron_path=macaron_path)
+        case BashScriptType.FILE:
+            parsed_content = parse_file(source_path, macaron_path=macaron_path)
+    bash_node = BashNode(
+        name,
+        node_type,
+        source_path,
+        parsed_step_obj=parsed_obj,
+        parsed_bash_obj=parsed_content,
+        node_id=node_id,
+        caller=caller,
+    )
+    caller_commands = parsed_content.get("commands", [])
 
     # Parse the bash script files called from the current script.
-    if recursive and repo_path:
+    if caller_commands and repo_path:
         for cmd in caller_commands:
             # Parse the scripts that end with `.sh`.
-            # We only parse recursively at depth 1, so don't set the recursive argument in parse_file().
             # TODO: parse Makefiles for bash commands.
-            if cmd[0] and cmd[0].endswith(".sh") and os.path.exists(os.path.join(repo_path, cmd[0])):
-                callee_commands = parse_file(os.path.join(repo_path, cmd[0])).get("commands", [])
-                if not callee_commands:
-                    continue
-
-                yield BashCommands(
-                    caller_path=os.path.join(working_dir, cmd[0]),
-                    CI_path=ci_file,
-                    CI_type=ci_type,
-                    commands=callee_commands,
-                    job_name=job_name,
-                    step_name=step_name,
-                )
+            if not cmd[0] or not cmd[0].endswith(".sh"):
+                continue
+
+            # Check for path traversal patterns before analyzing a bash file.
+            bash_file_path = os.path.realpath(os.path.join(repo_path, working_dir or "", cmd[0]))
```

Comment:
Should we set the `strict` parameter of `os.path.realpath` to True to detect symlink loop :thinking: ([docs](https://docs.python.org/3/library/os.path.html#os.path.realpath)).

Reply:
The reason I'm using `os.path.realpath` is to resolve the path. Path existence is checked separately.

Reply:
Right, I didn't mean to use the `strict` parameter to check for Path existence. I meant to use it for detecting symlink loops when we try to resolve the path. The exception raised with the path doesn't exist is a side effect of this `strict` parameter.

Reply:
Can you please give me a concrete example where the current implementation is not handling an issue correctly?

Reply:
After having a closer look, I think there is no issue with the current implementation.
The `strict` parameter if set to True will raise exception if there is a symlink loop.
However if `strict` parameter is set to False (by defaut) like what we have here, it can still handle symlink loop, it just not raise any exception. 
For example
```
>>> import os
>>> os.symlink("b", "b")
>>> os.path.realpath("b")
'.../b'
>>> os.path.realpath("b", strict=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<frozen posixpath>", line 416, in realpath
  File "<frozen posixpath>", line 476, in _joinrealpath
  File "<frozen posixpath>", line 471, in _joinrealpath
OSError: [Errno 40] Too many levels of symbolic links: 'b'
```
Therefore, I will close this conversation.